chore: Release 0.9.0 (#1969)
davidlattimore/wild: CHANGELOG.md

## 0.9.0

> The Wild repository moved to a github org, so is now at https://github.com/wild-linker/wild. The old

Reviewer: ```suggestion
The Wild repository moved to a GitHub org, so is now at https://github.com/wild-linker/wild. The old
```

URL should redirect to the new one.

A lot of work in this release was related to porting to other platforms. We did a large number of
refactorings to put ELF-specific behaviour behind traits. We've also started work on porting to Mac

> and Wasm. These aren't yet ready for use, but if you'd like to help out with porting, get in touch.

Reviewer: ```suggestion
refactorings to put ELF-specific behaviour behind traits. We've also started work on porting to Mach-O
and WebAssembly. These aren't yet ready for use, but if you'd like to help out with porting, get in touch.
```

are still a few known issues, but it's already working on a good range of programs. Pure Rust
projects generally don't use linker plugins, since Rust can do LTO within the compiler, so this is
mostly helpful for C, C++ or mixed language projects. Note that when using a linker plugin, link
times will be very slow.

Lots more linker-script features were implemented during this release.

### 🚀 Features

- Support `--dependency-file` (#1467)
- Add built-in symbol __executable_start (#1473)
- Allow specifying library search paths in version scripts and export lists (#1477)
- Add support for `--auxiliary` arg (#1488)
- Allow excluding symbols from specific libraries via `--exclude-libs` (#1494)
- Support `--enable-new-dtags` and `--disable-new-dtags` (#1507)
- Add flag to turn off version in .comment (#1504)
- Support `PROVIDE` and `PROVIDE_HIDDEN` linker script directives (#1525)
- Reject conflicting RISC-V ISA extensions (#1526)
- Add support for linker plugins (#1411)
- Add base to track byte reduction due to relaxation (#1527)
- Support RISC-V call relaxation (#1552)
- Define `_etext`, `__etext` and `_edata` special symbols (#1563)

> - Don't bail on --use-android-relr-tags and --pack-dyn-relocs=relr (#1566)

Reviewer 2: ```suggestion
- Don't bail on `--use-android-relr-tags` and `--pack-dyn-relocs=relr` (#1566)
```

- Add support for `--trace` (#1699)
- Evaluate linker script ASSERT commands (#1695)
- *(loongarch64)* Add allowed range and alignment by spec (#1734)
- Allow users of libwild to capture all warnings (#1756)
- *(AArch64)* Support 2 more relocations (#1760)
- Add support for generating static relocatables (#1698)
- Support bitwise, logical, and unary operators in linker script expressions (#1768)
- Implement ALIGNOF linker script function (#1770)
- Support `R_RISCV_ALIGN` relaxation (#1772)
- Support LOADADDR in linker scripts (#1780)
- Support K and M suffixes in linker script number literals (#1782)
- Support `-o/path` (without the space) (#1785)
- Add MEMORY block parsing and LENGTH/ORIGIN expression variants (#1792)
- Support `-z pack-relative-relocs` without actual packing (#1701)
- Support `--pack-dynamic-relocs=relr` (#1804)
- Implement ORIGIN and LENGTH evaluation in linker script expression eval (#1803)
- Generate a .symtab_shndx section when shnum > 65k (#1783)
- Handle output section header start addresses in linker scripts (#1850)
- Support `--use-android-relr-tags` (#1805)
- Implement range-extension thunks for aarch64 (#1847)

> - Select platform based on -flavor, arg0 then host (#1873)

Reviewer 2: ```suggestion
- Select platform based on `-flavor`, arg0 then host (#1873)
```

- Add -Ttext/-Tdata/-Tbss segment layout for SEGMENT_START support (#1877)
- Support SEGMENT_START function in Linker Script (#1851)

> - Support --compress-debug-sections (#1881)

Reviewer 2: ```suggestion
- Support `--compress-debug-sections` (#1881)
```

- Don't warn if unsupported `--pack-dyn-relocs` is overridden (#1874)
- Disable SFrame support by default (#1871)
- Emit error if attempting static link of dynamic object (#1886)
- Only keep RELRO_PADDING section when relro is enabled (#1892)
- Make --whole-archive work with linker plugins (#1900)
- *(jobserver)* ThreadPoolBuilder must use 1 thread with available_threads == 1 (#1904)
- Increase file limit when linker plugin is active (#1901)
- Implement plugin callsbacks required for thin LTO (#1902)
- Set file limit before we open input files (#1911)
- Bad allocation for versioned internal symbol (#1914)
- More robust code to get line numbers from debug info (#1923)
- Do not output SFrame section unless asked (#1916)
- Allow LTO to eliminate dead code (#1942)
- Handle missing version node for synthetic symbols (#1924)
- Fix resolution following linker scripts / synthetic symbols (#1953)
- Prevent GC of linker script symbol aliases (#1956)
- Load archive entries targeted by --defsym and linker script aliases (#1957)
- Emit error on missing resolution for symbol alias (#1961)

### 📚 Documentation

- Add code of conduct (#1630)
- Write up LLM use policy (#1618)

> - Change LTO status back to not-yet-supported (#1647)

Reviewer 3: Was this re-enabled by default in #1943?

- Add governance documentation (#1656)
- Update communication options section in `CONTRIBUTING.md` (#1676)
- Add linker script support matrix and feature tracking (#1827)
- Use icons for implementation status (#1839)
- Simplify usage with GCC 16.1 (-fuse-ld=wild) (#1883)
- *(README.md)* Adjust what's not supported yet (#1906)
- Add instructions for installing with Brew (#1913)

> ### 🕹️ Porting

Reviewer: How about sorting by platform order? Also, there's inconsistent notation regarding Mach-O (MachO, Macho and Mach-O). Wouldn't it be better to standardize this in the changelog?

- *(wasm)* Parse `linking` and `reloc.*` custom sections (#1918)
- *(MachO)* Emit code signature (#1919)
- *(wasm)* Add section and program segment mapping (#1926)
- *(Macho)* Introduce feature and smaller refactoring (#1929)
- *(wasm)* Implement object file section accessors (#1936)
- *(MachO)* Fat binary support (#1910)
- *(wasm)* Implement object file symbol accessors and `Symbol` trait (#1946)
- *(MachO)* Use args.output for CS identifier (#1949)
- *(Mach-O)* Include collected notes (#1812)
- *(wasm)* Support relocation encoding (#1964)
- Support running on wasi (#1723)

### 🎨 Styling

- Change C to 100 columns (#1928)
- Format TOML files and use taplo for checking (#1930)
- Add ymlfmt and format existing YAML files (#1652)

### 🛠️ Dev tooling

- Support passing a symbol ID to --sym-info (#1511)
- Fix save-dir when input filenames contain '=' (#1833)
- Save-dir now keeps at-files as separate files (#1834)
- Add some colour to --sym-info output (#1844)
- Use WILD_LOG environment variable rather than RUST_LOG (#1951)
- Make --sym-info show canonical targets regardless of order (#1952)
- Fix save-dir handling of response files (#1858)

### ⚖️ Linker Diff

> - Identify some relaxations that lld does (#1510)

Reviewer 2: ```suggestion
- Identify some relaxations that LLD does (#1510)
```
